internal/controller/replication.storage/volumegroupreplication_controller.go

```diff
@@ -18,36 +18,752 @@ package controller
+	if err != nil {
+		r.log.Error(err, "failed to list pvcs with the given selectors", "Selector", newSelector.String())
+		return nil, nil, "", err
+	}
+
+	// Update events if PVC is marked for deletion, but contains the pvc selector label of group
+	for _, pvc := range pvcList.Items {
+		if !pvc.DeletionTimestamp.IsZero() {
+			// PVC is marked for deletion, but not deleted because it is still a part of
+			// group using label selectors. Add an event to the PVC mentioning the same
+			msg := fmt.Sprintf("PersistentVolumeClaim is part of the group(%s/%s) using matching label selector. Remove label from PersistentVolumeClaim (%s/%s) to allow deletion",
+				instance.Namespace, instance.Name, pvc.Namespace, pvc.Name)
+			r.Recorder.Event(&pvc, "Warning", "PersistentVolumeClaimDeletionBlocked", msg)
+		}
+	}
+
+	pvHandlesList := []string{}
+	for _, pvc := range pvcList.Items {
+		pvName := pvc.Spec.VolumeName
+		pv := &corev1.PersistentVolume{}
+		err := r.Client.Get(r.ctx, types.NamespacedName{Name: pvName}, pv)
+		if err != nil {
+			r.log.Error(err, "failed to get pv for corresponding pvc", "PVC Name", pvc.Name)
+			return nil, nil, "", err
+		}
+		if pv.Spec.CSI == nil {
+			err = fmt.Errorf("pvc %s is not bound to a CSI PV", pvc.Name)
+			return nil, nil, "", err
+		}
+		if pv.Spec.CSI.Driver != vgrClass.Spec.Provisioner {
+			err = fmt.Errorf("driver (%s) of PV for PVC %s is different than the VolumeGroupReplicationClass driver (%s)", pvc.Name, pv.Spec.CSI.Driver, vgrClass.Spec.Provisioner)
+			return nil, nil, "", err
+		}
+		pvHandlesList = append(pvHandlesList, pv.Spec.CSI.VolumeHandle)
+	}
+
+	// Sort the pvHandles list and then pass, because reflect.DeepEqual checks for positional equality
+	slices.Sort(pvHandlesList)
+
+	return pvcList.Items, pvHandlesList, newSelector.String(), nil
+}
+
+func (r *VolumeGroupReplicationReconciler) updateFinalizerAndAnnotationOnPVCs(vgr *replicationv1alpha1.VolumeGroupReplication,
+	pvcList []corev1.PersistentVolumeClaim) ([]corev1.LocalObjectReference, error) {
+
+	// Create a set of old PVC names
+	oldPVCs := make(map[string]struct{}, len(vgr.Status.PersistentVolumeClaimsRefList))
+	for _, pvc := range vgr.Status.PersistentVolumeClaimsRefList {
+		oldPVCs[pvc.Name] = struct{}{}
+	}
+
+	// Track PVCs to add and remove
+	toAddPVCs := make([]corev1.PersistentVolumeClaim, 0)
+	toRemovePVCs := make([]corev1.PersistentVolumeClaim, 0)
+
+	// Create a set for current PVC names to determine removals
+	currentPVCs := make(map[string]struct{}, len(pvcList))
+	pvcRefList := []corev1.LocalObjectReference{}
+	for _, pvc := range pvcList {
+		currentPVCs[pvc.Name] = struct{}{}
+		pvcRefList = append(pvcRefList, corev1.LocalObjectReference{
+			Name: pvc.Name,
+		})
+
+		// If it's not in oldPVCs, it's an addition
+		if _, exists := oldPVCs[pvc.Name]; !exists {
+			toAddPVCs = append(toAddPVCs, pvc)
+		}
+	}
+
+	// Identify PVCs that should be removed
+	for name := range oldPVCs {
```

Comment:
In other for loops you use `_` for the index, here you don't. Please follow the same coding style everywhere.

Reply:
Actually, this is a map that we are iterating over.
Though, I have updated it use `_` style for values, as those are unused.